src/curl.c

```diff
@@ -128,18 +128,20 @@ static size_t rcurl_read(void *target, size_t sz, size_t ni, Rconnection con) {
   request *req = (request*) con->private;
   size_t req_size = sz * ni;
 
+  if (!con->blocking || req->partial) {
+    // If the connection is non-blocking (or using curl_fetch_stream()), always
+    // fetch data first (as likely to be called when file descriptor has been
+    // indicated ready for reading). Then just return what's available without
+    // waiting.
+    fetchdata(req);
+    size_t total_size = pop(target, req_size, req);
+    con->incomplete = req->has_more || req->size;
+    return total_size;
+  }
+
```

Comment:
What about the part of the comment below that says we want to flush already-received bytes before a connection error? Could this potentially change that behavior?

Was the old code causing problems with `later_fd`, or is the benefit of this refactoring more theoretical?

Reply:
It was getting stuck before (not just theoretical), but I rebased to the latest curl version and it seems no longer the case - but the git blame shows no change in that code for 10 years, so I'll have to look at it in a bit more depth.

The flush bytes before connection error I think will have to be a necessary casualty of this change - basically the fd being ready means we *need* to call curl to perform what's necessary (according to its docs). At the end of the day, if the connection is shut down mid-way, does it matter if we get a few more bytes vs not? Jeroen will probably have more context for this.